File: PRs/5923 Dereferencing as reassignment target/reassignments_lhs_rhs_evaluation_order/play.cpp
#include <iostream>

int index(char* s, int i) {
    std::cout << s << std::endl;
    return i;
}

int inc_i(int& i) {
    i += 1;
    return i;
}

int main()
{
    int array[3] = { 0, 0, 0 };
    
    array[index("LHS", 0)] = index("RHS", 11);
    
    std::cout << array[0] << std::endl;
    
    std::cout << "-------------------------" << std::endl;
    
    int i = 0;
    array[inc_i(i)] = inc_i(i);
    std::cout << array[0] << std::endl;
    std::cout << array[1] << std::endl;
    std::cout << array[2] << std::endl;

    return 0;
}


// LHS
// RHS
// 11
// -------------------------
// 11
// 2
// 0

File: PRs/5923 Dereferencing as reassignment target/reassignments_lhs_rhs_evaluation_order/play.rs
fn main() {
    let mut array = [0, 0, 0];
    
    array[index("LHS", 0)] = index("RHS", 11);
    
    println!("{}", array[0]);
    
    println!("-------------------------");
    
    let mut i = 0;
    array[inc_i(&mut i)] = inc_i(&mut i);
    println!("{}", array[0]);
    println!("{}", array[1]);
    println!("{}", array[2]);
}

fn index(s: &str, i: usize) -> usize {
    println!("{s}");
    i
}

fn inc_i(i: &mut usize) -> usize {
    *i += 1;
    *i
}

// RHS
// LHS
// 11
// -------------------------
// 11
// 0
// 1
